### ISSUE TITLE
[Scaffolder] Cannot open file when using Template Editor 

### ISSUE THREAD

**Issue by 18521542** [backstage/backstage#12075]
## Expected Behavior
When choosing `Load Template Directory` or `Edit Template Form` of Scaffolder's Template Editor, it can work like this PR mentioned: https://github.com/backstage/backstage/pull/11173

## Actual Behavior
- I have bumped backstage up from v1.0.0 to v1.3.0 but getting error when choosing `Load Template Directory` or `Edit Template Form`.
-  The new functionalities of Scaffolder get an error when trying to open the folder directory or load the exist template form. 
- The error is 
![image](https://user-images.githubusercontent.com/56584175/173983321-7f5fba30-be21-4739-bc1d-380ee5aa14cc.png)
- The error also happens with the app created by `create-app@latest` from `backstage-cli`
## Steps to Reproduce

1. Use `backstage-cli` to create the latest app
2. Go to create tab and click on the Template Editor
![image](https://user-images.githubusercontent.com/56584175/173986729-e3d89b5e-b847-4c43-b478-d60504bb2652.png)
3. Choose any option and you should see the error

## Context
If this fixed, I can improve my development experience with Software Template 

## Your Environment

- My OS: window

```text

```

**Comment by marcus-crane** [comment]
I've been having the same issue with multiple instances of `@codemirror/state`.

I know that Codemirror is used as the editor in the GraphQL plugin so I removed that thinking it might be the culprit but no luck.

Perhaps other plugins make use of Codemirror too? I haven't tracked any down just yet

**Comment by marcus-crane** [comment]
Ah ok, I think I've figured out what has gone wrong and why it suddenly started breaking despite what I'm sure was extensive testing before v1.3.0 released.

I'll do a write up in a sec but for now, you should be able to temporarily fix this issue by adding the following [resolution](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/) to your `package.json`:

```json
"resolutions": {
    "@uiw/react-codemirror": "4.7.0"
}
```

**Comment by marcus-crane** [comment]
**tl;dr**: `@uiw/react-codemirror` [recently introduced a direct dependency on `codemirror`](https://github.com/uiwjs/react-codemirror/commit/5cef465493407bf980e126956c47a0c9e4774573#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R66), resulting in a second instance of `@codemirror/state` that breaks the scaffolder editor.

---

The first instance of `@codemirror/state` is used by the API Docs plugin to render GraphQL schema definitions.

For context, [Codemirror](https://codemirror.net/) is a web text editor library, in the same vein as [Monaco](https://microsoft.github.io/monaco-editor/) which powers VS Code.

It's used in two plugins to my knowledge: [scaffolder](https://github.com/backstage/backstage/blob/master/plugins/scaffolder/package.json#L62) and [api-docs](https://github.com/backstage/backstage/blob/master/plugins/api-docs/package.json#L46)

Scaffolder uses `codemirror`by way of `@uiw/react-codemirror`, a React wrapper for Codemirror and API Docs uses it indirectly via `graphiql`

More specifically, `graphiql` is [used by the API Docs plugin](https://github.com/backstage/backstage/blob/ac511f98c433f0ed8483a6713e5110333ce11738/plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx#L19) to render GraphQL schemas as mentioned [in the plugin page](https://github.com/backstage/backstage/blob/master/plugins/api-docs/README.md#api-documentation).

Scaffolder, as we've know from this bug, newly uses `codemirror` to render an editor that lets you live edit templates.

We can see the resolution paths of both of these plugins like so:

```shell
$ yarn why codemirror
yarn why v1.22.19
[1/4] 🤔  Why do we have the module "codemirror"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "codemirror@6.0.0"
info Reasons this module exists
   - "_project_#app#@backstage#plugin-scaffolder#@uiw#react-codemirror" depends on it
   - Hoisted from "_project_#app#@backstage#plugin-scaffolder#@uiw#react-codemirror#codemirror"
info Disk size without dependencies: "4.07MB"
info Disk size with unique dependencies: "6.13MB"
info Disk size with transitive dependencies: "6.83MB"
info Number of shared dependencies: 12
=> Found "@graphiql/react#codemirror@5.65.5"
info This module exists because "_project_#app#@backstage#plugin-api-docs#graphiql#@graphiql#react" depends on it.
✨  Done in 1.02s.
```

We have two instances of `codemirror`, or more specifically `@codemirror/state`:

`v5.65.5` pulled in from `@backstage/plugin-apidocs` -> `graphiql` -> [`@graphiql/react`](https://github.com/graphql/graphiql/blob/main/packages/graphiql-react/package.json) -> [`codemirror`](https://github.com/graphql/graphiql/blob/main/packages/graphiql-react/package.json#L34)
`v6.0.0` pulled in from `@backstage/plugin-scaffolder` -> `@uiw/react-codemirror` -> [`codemirror`](https://github.com/uiwjs/react-codemirror/blob/master/package.json#L66)

That explains why but it doesnt' answer the question of how this was never noticed before!

Being a runtime crash, there's nothing obvious at build time that'll tell you this is broken at all.

Looking at the timing, [v1.3.0-next.2](https://github.com/backstage/backstage/releases/tag/v1.3.0-next.2) was introduced 10 days ago (at the time of writing) while [v1.3.0](https://github.com/backstage/backstage/releases/tag/v1.3.0) was officially released 3 days ago.

In a case of miserable timing, [v4.8.0](https://github.com/uiwjs/react-codemirror/commit/5cef465493407bf980e126956c47a0c9e4774573#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R66) of Codemirror was released 8 days ago (during that window), replacing some indirect dependencies for a direct dependency on Codemirror itself.

Given that `api-docs` had a caret (`^`) dependency on `@uiw/react-codemirror": "^4.7.0"`, this bump to `4.8.0` was implicitly pulled in for fresh installs, silently introducing a second instance of `@codemirror/state` and breaking the scaffolder presumably because API Docs had already loaded its instance first.

Anyway, the resolution fix suggested works because it downgrades `@uiw/react-codemirror` to the last version before this direct dependency on codemirror was introduced.

It may be the case that having the same versions of codemirror under the hood for `graphiql` and `@uiw/react-codemirror` would get resolved to a single instance of `@codemirror/state` if that is even possible but at face value, I'm not sure if that would imply that state would be shared between all instances of the text editor? Surely not.

Anyway, hopefully that should give enough background to come up with a proper solution longer term.

**Comment by 18521542** [comment]
Yeah, so I added `"resolutions": {
"@uiw/react-codemirror": "4.7.0"
}` to my `package.json` 
=> It works fine with the `Edit Template Form` but I must wait for a VERY long time to see the directory loaded when choosing `Load Template Directory`. Did you have the same issue? @marcus-crane 
I'm using window 10

**Comment by marcus-crane** [comment]
Ah, I haven't tried the "Edit Template Form" feature just yet but I'll give it a spin later on

**Comment by 18521542** [comment]
And also cannot use `try it` button of the Template Editor 
![image](https://user-images.githubusercontent.com/56584175/174221108-980c57a5-7c9a-4207-a4ab-6e2b0dc987d8.png)

**Comment by regicsolutions** [comment]
Thanks @marcus-crane this also fixed my issue as well! i am also facing the same issue as @18521542 with the 413 error:
![2022-06-17_8-58-44](https://user-images.githubusercontent.com/6157895/174309718-8c23f152-cf35-4a9d-ab09-03e5612a25df.png)
r

**Comment by goenning** [comment]
@regicsolutions I had the same issue and I fixed it by adding this to backend/index.ts
```
  apiRouter.use(bodyParser({ limit: '2mb' }));
```

Apparently express has a default limit of 100kb for the payload, and our template was 1.5Mb, not sure why yet, will have to investigate

**Comment by Rugvip** [comment]
Will try to get codemirror fix in for the prerelease tomorrow.

The size limit is a bit trickier. There is a cutoff at which point the file contents gets replaced with placeholder text, but it's configured to still allow somewhat large files. Could probably reduce the limit a fair bit though, but it does mean some large text files might not be visible.

**Comment by marcus-crane** [comment]
🏃💨

**Comment by francescoben** [comment]
> @regicsolutions I had the same issue and I fixed it by adding this to backend/index.ts
> 
> ```
>   apiRouter.use(bodyParser({ limit: '2mb' }));
> ```
> 
> Apparently express has a default limit of 100kb for the payload, and our template was 1.5Mb, not sure why yet, will have to investigate

hi @goenning, I'm facing the "payload too large" issue because my template has one image field and its value is the very long base64 string of the image. I tried to add this line in my backend/index.ts didn't work, could you tell me where to put this line exactly?

**Comment by CarlosPet** [comment]
> > @regicsolutions I had the same issue and I fixed it by adding this to backend/index.ts
> > ```
> >   apiRouter.use(bodyParser({ limit: '2mb' }));
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Apparently express has a default limit of 100kb for the payload, and our template was 1.5Mb, not sure why yet, will have to investigate
> 
> hi @goenning, I'm facing the "payload too large" issue because my template has one image field and its value is the very long base64 string of the image. I tried to add this line in my backend/index.ts didn't work, could you tell me where to put this line exactly?

I tried, and like the other comment here, Changing that parameter didn't solve my issue initially.
To anyone who comes across this page. be sure to declare the route as the first option in the code. otherwise it will not work. so it will look like this:

```Typescript
  const apiRouter = Router();
  apiRouter.use(bodyParser({ limit: '50mb' })); // add this right after declaring the apiRouter
  apiRouter.use('/catalog', await catalog(catalogEnv));
  apiRouter.use('/scaffolder', await scaffolder(scaffolderEnv));
  // ... other routes
```
it works for me now
